### PR TITLE
Proof system refactor

### PR DESCRIPTION
**Pull request by dark64**
Work in progress

### CODE REVIEW

File: zokrates_cli/tests/contract/test.js [hunk 1]
```diff
@@ -64,6 +64,7 @@ let jsonInterface = JSON.parse(solc.compile(jsonContractSource));
 
     function makeTransaction(account, correct) {
         let proof = getProof(correct);
+        console.log(proof)
```
Comment (Schaeff) [inline zokrates_cli/tests/contract/test.js:67]: should this stay?

File: zokrates_cli/src/bin.rs [hunk 2]
```diff
@@ -691,76 +740,111 @@ fn cli() -> Result<(), String> {
...
-                constants::PGHR13 => match curve {
-                    constants::BN128 => cli_export_verifier::<Bn128Field, PGHR13>(sub_matches)?,
-                    _ => unimplemented!(),
-                },
+            match backend {
+                constants::BELLMAN => match proof_system {
+                    constants::G16 => match curve {
+                        constants::BN128 => cli_export_verifier::<Bn128Field, G16>(sub_matches),
+                        constants::BLS12_381 => cli_export_verifier::<Bls12Field, G16>(sub_matches),
+                        _ => unimplemented!(),
```
Comment (Schaeff) [inline zokrates_cli/src/bin.rs:784]: I guess this is unreachable as clap checked the options before

File: zokrates_cli/src/bin.rs [hunk 3]
```diff
@@ -691,76 +740,111 @@ fn cli() -> Result<(), String> {
...
+                }?,
                 #[cfg(feature = "libsnark")]
-                constants::GM17 => match curve {
-                    constants::BN128 => cli_export_verifier::<Bn128Field, GM17>(sub_matches)?,
-                    _ => unimplemented!(),
-                },
+                constants::LIBSNARK => match proof_system {
+                    constants::PGHR13 => match curve {
+                        constants::BN128 => cli_export_verifier::<Bn128Field, PGHR13>(sub_matches),
+                        _ => unimplemented!(),
```
Comment (Schaeff) [inline zokrates_cli/src/bin.rs:795]: this should return a clean error

File: zokrates_cli/src/bin.rs [hunk 4]
```diff
@@ -691,76 +740,111 @@ fn cli() -> Result<(), String> {
...
+                constants::LIBSNARK => match proof_system {
+                    constants::PGHR13 => match prog {
+                        ProgEnum::Bn128Program(p) => {
+                            cli_generate_proof::<_, PGHR13>(p, sub_matches)
+                        }
+                        _ => unimplemented!(),
+                    },
+                    constants::GM17 => match prog {
+                        ProgEnum::Bn128Program(p) => cli_generate_proof::<_, GM17>(p, sub_matches),
+                        _ => unimplemented!(),
```
Comment (Schaeff) [inline zokrates_cli/src/bin.rs:841]: same here

File: zokrates_cli/src/bin.rs [hunk 5]
```diff
@@ -803,24 +887,36 @@ fn cli() -> Result<(), String> {
...
-                constants::PGHR13 => match curve {
-                    constants::BN128 => cli_verify::<Bn128Field, PGHR13>(sub_matches)?,
-                    _ => unimplemented!(),
-                },
+            match backend {
+                constants::BELLMAN => match proof_system {
+                    constants::G16 => match curve {
+                        constants::BN128 => cli_verify::<Bn128Field, G16>(sub_matches),
+                        constants::BLS12_381 => cli_verify::<Bls12Field, G16>(sub_matches),
+                        _ => unimplemented!(),
```
Comment (Schaeff) [inline zokrates_cli/src/bin.rs:898]: same here

File: zokrates_cli/tests/contract/test.js (hunk 1 above)
Comment (dark64) [reply zokrates_cli/tests/contract/test.js]: good catch, forgot to remove

File: zokrates_cli/src/helpers.rs [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+pub enum Backend {
+    Bellman,
+    #[cfg(feature = "libsnark")]
+    Libsnark,
+}
+
+pub enum ProvingScheme {
+    G16,
+    #[cfg(feature = "libsnark")]
```
Comment (Schaeff) [inline zokrates_cli/src/helpers.rs:18]: I see why this is like this but ideally we would want to decouple backend and proving scheme and keep all backends in there.

**Pull request merged**